### ISSUE TITLE
Overriding `panic` function causes compiler crash

### ISSUE THREAD

**Issue by hqnna** [ziglang/zig#16327]
## Zig Version

0.11.0-dev.3934+ba6e5e65a

### Reproduction Steps

1. Create a file named `example.zig` with the following content
```zig
const std = @import("std");

pub fn panic(_: []const u8, _: ?*std.builtin.StackTrace, _: ?usize) noreturn {}

pub fn main() !void {}
```
2. Attempt to build the file with the following
```
zig build-exe example.zig
```
### Observed Behavior
The compiler crashes with a segmentation fault error.
```
zsh: segmentation fault (core dumped)  zig build-exe test.zig
```

## Expected Behavior

The compiler should not crash when attempting to override the `panic` function.

**Comment by notcancername** [comment]
It gives this error with my debug build:
```
repr.zig:3:69: error: function declared 'noreturn' implicitly returns
pub fn panic(_: []const u8, _: ?*std.builtin.StackTrace, _: ?usize) noreturn {
                                                                    ^~~~~~~~
repr.zig:5:1: note: control flow reaches end of body here
}
^
referenced by:
    panic: /home/user/src/zig/zig-out/lib/zig/std/builtin.zig:725:9
    remaining reference traces hidden; use '-freference-trace' to see all reference traces
```
Using this:

```
pub fn panic(_: []const u8, _: ?*std.builtin.StackTrace, _: ?usize) noreturn {
    std.os.abort();
}
```
Works.

**Comment by ianprime0509** [comment]
The issue here is in attempting to format the reference trace for the error ("function declared 'noreturn' implicitly returns"). In this case, the first element in the reference trace is the panic call, which currently has an `unneeded` source location, causing unreachable code to be reached in this call to `span`: https://github.com/ziglang/zig/blob/e05c242cd89d03ae67fa7b8d5fc33fb5dc42dbe3/src/Compilation.zig#L2825

This patch fixes the crash by assigning a proper source location to panics:

<details>
<summary>Patch</summary>

```diff
--- a/src/Sema.zig
+++ b/src/Sema.zig
@@ -5395,7 +5395,7 @@ fn zirPanic(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError!Zir.I
     if (block.is_comptime) {
         return sema.fail(block, src, "encountered @panic at comptime", .{});
     }
-    try sema.panicWithMsg(block, msg_inst);
+    try sema.panicWithMsg(block, src, msg_inst);
     return always_noreturn;
 }
 
@@ -24973,7 +24973,7 @@ fn addSafetyCheckExtra(
     parent_block.instructions.appendAssumeCapacity(block_inst);
 }
 
-fn panicWithMsg(sema: *Sema, block: *Block, msg_inst: Air.Inst.Ref) !void {
+fn panicWithMsg(sema: *Sema, block: *Block, src: LazySrcLoc, msg_inst: Air.Inst.Ref) !void {
     const mod = sema.mod;
 
     if (!mod.backendSupportsFeature(.panic_fn)) {
@@ -24984,7 +24984,7 @@ fn panicWithMsg(sema: *Sema, block: *Block, msg_inst: Air.Inst.Ref) !void {
     try sema.prepareSimplePanic(block);
 
     const panic_func = mod.funcPtrUnwrap(mod.panic_func_index).?;
-    const panic_fn = try sema.analyzeDeclVal(block, .unneeded, panic_func.owner_decl);
+    const panic_fn = try sema.analyzeDeclVal(block, src, panic_func.owner_decl);
     const null_stack_trace = try sema.addConstant(mod.null_stack_trace.toValue());
 
     const opt_usize_ty = try mod.optionalType(.usize_type);
@@ -25147,7 +25147,8 @@ fn safetyCheckFormatted(
 fn safetyPanic(sema: *Sema, block: *Block, panic_id: Module.PanicId) CompileError!void {
     const msg_decl_index = try sema.preparePanicId(block, panic_id);
     const msg_inst = try sema.analyzeDeclVal(block, sema.src, msg_decl_index);
-    try sema.panicWithMsg(block, msg_inst);
+    const decl = sema.mod.declPtr(block.src_decl);
+    try sema.panicWithMsg(block, decl.srcLoc(sema.mod).lazy, msg_inst);
 }
 
 fn emitBackwardBranch(sema: *Sema, block: *Block, src: LazySrcLoc) !void {
```

</details>

However, I don't entirely feel like this is the correct location to use for safety check panics (it's too broad). For example, here's the error message for the original example after this change:

```
test.zig:3:69: error: function declared 'noreturn' implicitly returns
pub fn panic(_: []const u8, _: ?*std.builtin.StackTrace, _: ?usize) noreturn {}
                                                                    ^~~~~~~~
test.zig:3:79: note: control flow reaches end of body here
pub fn panic(_: []const u8, _: ?*std.builtin.StackTrace, _: ?usize) noreturn {}
                                                                              ^
referenced by:
    posixCallMainAndExit: /home/ian/src/zig/lib/std/start.zig:393:1
    _start: /home/ian/src/zig/lib/std/start.zig:367:26
```

This is referencing the fact that a safety check is emitted _somewhere_ in `posixCallMainAndExit`, but it doesn't say where. (This is also why the original example works fine with `-OReleaseFast`, because no safety check is emitted, and hence no call to panic)

If anyone has any suggestions on this, I'd be happy to apply them and make a PR for this.

**Comment by xdBronch** [comment]
#16798 seems to have fixed this

**Comment by jacobly0** [comment]
Fixed by 8f3ccbbe367bea66d7f0f364a957870eb2cc95a0.
